packages/devtools-evm-hardhat/src/omnigraph/coordinates.ts

```diff
@@ -18,6 +18,75 @@ export const omniDeploymentToContract = ({ eid, deployment }: OmniDeployment): O
     contract: new Contract(deployment.address, deployment.abi),
 })
 
+export const createAbiFactory = (environmentFactory = createGetHreByEid()): AbiFactoryHardhat => {
+    return pMemoize(async ({ eid, address, contractName }) => {
+        const env = await environmentFactory(eid)
+        assertHardhatDeploy(env)
+
+        const networkLabel = `${formatEid(eid)} (${env.network.name})`
+        const logger = createModuleLogger(`Contract factory @ ${networkLabel}`)
+
+        // If we have both the contract name & address, we go off artifacts
+        if (contractName != null && address != null) {
+            logger.verbose(`Looking for contract ${contractName} on address ${address} in artifacts`)
+
+            const artifact = await env.deployments.getArtifact(contractName).catch((error) => {
+                logger.verbose(`Failed to load artifact for contract ${contractName} on address ${address}: ${error}`)
+                logger.verbose(`Will search for the contract by its address only`)
+            })
+
+            if (artifact != null) {
+                return artifact.abi
+            }
+        }
+
+        // If we have the contract name but no address, we need to get it from the deployments by name
+        if (contractName != null && address == null) {
+            logger.verbose(`Looking for contract ${contractName} in deployments`)
+
+            const deployment = await env.deployments.getOrNull(contractName)
+            assert(deployment != null, `Could not find a deployment for contract '${contractName}' on ${networkLabel}`)
+
+            return deployment.abi
+        }
+
+        // And if we only have the address, we need to go get it from deployments by address
```

Comment:
Doesn't this still enforce that you have the hardhat-deploy JSON artifact in the deployments folder?

Wouldn't a true decoupling mean we provide the ABI for the base OApp standard and just check that the address has these ABI methods?

Reply:
Decoupling will not mean that we lose all the hardhat deploy functionality, we'll still provide it for people to use and they can opt in if they want. For example Stargate still benefits from hardhat-deploy functionality